ICE: None in compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
pub trait TraitWAssocConst {
  const A:   usize;
}
pub struct Demo {}

impl TraitWAssocConst for  impl Demo {
  pubconst A: str = 32;
}

fn foo<A: TraitWAssocConst<A=32>>() {
  foo::<Demo>()();
}
//~^ ERROR associated const equality

fn main<A: TraitWAssocConst<A=32>>() {
  foo::<Demo>();
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (53e4b9dd7 2022-12-04)
binary: rustc
commit-hash: 53e4b9dd74c29cc9308b8d0f10facac70bb101a7
commit-date: 2022-12-04
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

### Error output

```
error: expected one of `!` or `::`, found `A`
 --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:7:12
  |
6 | impl TraitWAssocConst for  impl Demo {
  |                                      - while parsing this item list starting here
7 |   pubconst A: str = 32;
  |            ^ expected one of `!` or `::`
8 | }
  | - the item list ends here

error[E0404]: expected trait, found struct `Demo`
 --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:6:33
  |
6 | impl TraitWAssocConst for  impl Demo {
  |                                 ^^^^ not a trait

error[E0658]: associated const equality is incomplete
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:10:28
   |
10 | fn foo<A: TraitWAssocConst<A=32>>() {
   |                            ^^^^
   |
   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable

error[E0658]: associated const equality is incomplete
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:15:29
   |
15 | fn main<A: TraitWAssocConst<A=32>>() {
   |                             ^^^^
   |
   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable

error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in type
 --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:6:28
  |
6 | impl TraitWAssocConst for  impl Demo {
  |                            ^^^^^^^^^

error[E0277]: the trait bound `Demo: TraitWAssocConst` is not satisfied
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:11:9
   |
11 |   foo::<Demo>()();
   |         ^^^^ the trait `TraitWAssocConst` is not implemented for `Demo`
   |
note: required by a bound in `foo`
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:10:11
   |
10 | fn foo<A: TraitWAssocConst<A=32>>() {
   |           ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `foo`
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1610:35
stack backtrace:
   0:     0x7f48a3b662da - std::backtrace_rs::backtrace::libunwind::trace::h217eb04aed2c45de
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f48a3b662da - std::backtrace_rs::backtrace::trace_unsynchronized::h8d7ebe2490fc8887
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f48a3b662da - std::sys_common::backtrace::_print_fmt::hf5be7add3e3c061b
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f48a3b662da - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc3f0b6fc09e1404b
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f48a3bc8ece - core::fmt::write::h21bd73c842f4e089
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/fmt/mod.rs:1208:17
   5:     0x7f48a3b56795 - std::io::Write::write_fmt::hc308562d08004f3e
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/io/mod.rs:1682:15
   6:     0x7f48a3b660a5 - std::sys_common::backtrace::_print::h35dbc8fac7f76168
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f48a3b660a5 - std::sys_common::backtrace::print::h73d53422718fe2c6
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f48a3b68def - std::panicking::default_hook::{{closure}}::h48456de670413439
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:267:22
   9:     0x7f48a3b68b2b - std::panicking::default_hook::h133845cf05f47840
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:286:9
  10:     0x7f48a3b6960c - std::panicking::rust_panic_with_hook::hb428246b31e531d9
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:688:13
  11:     0x7f48a3b69362 - std::panicking::begin_panic_handler::{{closure}}::ha15618c185ebb8be
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:577:13
  12:     0x7f48a3b6678c - std::sys_common::backtrace::__rust_end_short_backtrace::ha22c9026619d93a1
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:137:18
  13:     0x7f48a3b690b2 - rust_begin_unwind
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:575:5
  14:     0x7f48a3bc58b3 - core::panicking::panic_fmt::hee7675342c7d8922
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/panicking.rs:64:14
  15:     0x7f48a3bc598d - core::panicking::panic::h934e7549db567587
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/panicking.rs:111:5
  16:     0x7f48a7ae26a5 - <rustc_infer[367541f3d053723]::infer::InferCtxt>::probe::<(), <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::InferCtxtPrivExt>::report_projection_error::{closure#0}>
  17:     0x7f48a7b22c6d - <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::InferCtxtPrivExt>::report_projection_error
  18:     0x7f48a7b16d94 - <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::TypeErrCtxtExt>::report_fulfillment_errors
  19:     0x7f48a5070188 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_argument_types
  20:     0x7f48a52483f9 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_call
  21:     0x7f48a51a23ce - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  22:     0x7f48a523c599 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_call
  23:     0x7f48a51a23ce - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  24:     0x7f48a51f57a3 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_block_with_expected
  25:     0x7f48a51a27cf - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  26:     0x7f48a518deb3 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_return_expr
  27:     0x7f48a518365f - rustc_hir_typeck[62f7ea54373b9441]::check::check_fn
  28:     0x7f48a516e47b - rustc_hir_typeck[62f7ea54373b9441]::typeck
  29:     0x7f48a554d4a2 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::VecCache<rustc_span[323ba4f8bef90746]::def_id::LocalDefId, &rustc_middle[962fafeac42f2169]::ty::context::TypeckResults>>
  30:     0x7f48a6636c05 - rustc_data_structures[13617afd267f24a6]::sync::par_for_each_in::<&[rustc_span[323ba4f8bef90746]::def_id::LocalDefId], <rustc_middle[962fafeac42f2169]::hir::map::Map>::par_body_owners<rustc_hir_typeck[62f7ea54373b9441]::typeck_item_bodies::{closure#0}>::{closure#0}>
  31:     0x7f48a6636943 - rustc_hir_typeck[62f7ea54373b9441]::typeck_item_bodies
  32:     0x7f48a65552d7 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::DefaultCache<(), ()>>
  33:     0x7f48a6554fe0 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::get_query::<rustc_query_impl[3365ff5720cc62]::queries::typeck_item_bodies, rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt>
  34:     0x7f48a52e2c02 - <rustc_session[f586d2b4d440cc8a]::session::Session>::time::<(), rustc_hir_analysis[4b631a2ab94469f]::check_crate::{closure#7}>
  35:     0x7f48a52e19dc - rustc_hir_analysis[4b631a2ab94469f]::check_crate
  36:     0x7f48a52e15fb - rustc_interface[e2cb08752aecdf39]::passes::analysis
  37:     0x7f48a669ec0a - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::DefaultCache<(), core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>>
  38:     0x7f48a669e900 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::get_query::<rustc_query_impl[3365ff5720cc62]::queries::analysis, rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt>
  39:     0x7f48a603a2ab - <rustc_interface[e2cb08752aecdf39]::passes::QueryContext>::enter::<rustc_driver[b218747198152f07]::run_compiler::{closure#1}::{closure#2}::{closure#2}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  40:     0x7f48a6031c5e - <rustc_interface[e2cb08752aecdf39]::interface::Compiler>::enter::<rustc_driver[b218747198152f07]::run_compiler::{closure#1}::{closure#2}, core[e1ff35b71d1701d9]::result::Result<core[e1ff35b71d1701d9]::option::Option<rustc_interface[e2cb08752aecdf39]::queries::Linker>, rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  41:     0x7f48a602cc78 - rustc_span[323ba4f8bef90746]::with_source_map::<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  42:     0x7f48a602c765 - <scoped_tls[b7ce8b7e4286e02d]::ScopedKey<rustc_span[323ba4f8bef90746]::SessionGlobals>>::set::<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  43:     0x7f48a602bd52 - std[2f49e04a49785d41]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e2cb08752aecdf39]::util::run_in_thread_pool_with_globals<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  44:     0x7f48a67a5458 - <<std[2f49e04a49785d41]::thread::Builder>::spawn_unchecked_<rustc_interface[e2cb08752aecdf39]::util::run_in_thread_pool_with_globals<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#1} as core[e1ff35b71d1701d9]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  45:     0x7f48a7cb5df3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5d679a3e0eef548a
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/alloc/src/boxed.rs:2000:9
  46:     0x7f48a7cb5df3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5129455b92134839
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/alloc/src/boxed.rs:2000:9
  47:     0x7f48a7cb5df3 - std::sys::unix::thread::Thread::new::thread_start::haa6aecc225cb6cab
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys/unix/thread.rs:108:17
  48:     0x7f48a39068fd - <unknown>
  49:     0x7f48a3988a60 - <unknown>
  50:                0x0 - <unknown>
error[E0277]: the trait bound `Demo: TraitWAssocConst` is not satisfied
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:16:9
   |
16 |   foo::<Demo>();
   |         ^^^^ the trait `TraitWAssocConst` is not implemented for `Demo`
   |
note: required by a bound in `foo`
  --> ./0371a51e32b8e2e68e80645b20197ccad375a913.rs:10:11
   |
10 | fn foo<A: TraitWAssocConst<A=32>>() {
   |           ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `foo`

thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1610:35
stack backtrace:
   0:     0x7f48a3b662da - std::backtrace_rs::backtrace::libunwind::trace::h217eb04aed2c45de
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f48a3b662da - std::backtrace_rs::backtrace::trace_unsynchronized::h8d7ebe2490fc8887
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f48a3b662da - std::sys_common::backtrace::_print_fmt::hf5be7add3e3c061b
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f48a3b662da - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc3f0b6fc09e1404b
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f48a3bc8ece - core::fmt::write::h21bd73c842f4e089
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/fmt/mod.rs:1208:17
   5:     0x7f48a3b56795 - std::io::Write::write_fmt::hc308562d08004f3e
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/io/mod.rs:1682:15
   6:     0x7f48a3b660a5 - std::sys_common::backtrace::_print::h35dbc8fac7f76168
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f48a3b660a5 - std::sys_common::backtrace::print::h73d53422718fe2c6
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f48a3b68def - std::panicking::default_hook::{{closure}}::h48456de670413439
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:267:22
   9:     0x7f48a3b68b2b - std::panicking::default_hook::h133845cf05f47840
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:286:9
  10:     0x7f48a3b6960c - std::panicking::rust_panic_with_hook::hb428246b31e531d9
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:688:13
  11:     0x7f48a3b69362 - std::panicking::begin_panic_handler::{{closure}}::ha15618c185ebb8be
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:577:13
  12:     0x7f48a3b6678c - std::sys_common::backtrace::__rust_end_short_backtrace::ha22c9026619d93a1
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys_common/backtrace.rs:137:18
  13:     0x7f48a3b690b2 - rust_begin_unwind
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/panicking.rs:575:5
  14:     0x7f48a3bc58b3 - core::panicking::panic_fmt::hee7675342c7d8922
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/panicking.rs:64:14
  15:     0x7f48a3bc598d - core::panicking::panic::h934e7549db567587
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/core/src/panicking.rs:111:5
  16:     0x7f48a7ae26a5 - <rustc_infer[367541f3d053723]::infer::InferCtxt>::probe::<(), <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::InferCtxtPrivExt>::report_projection_error::{closure#0}>
  17:     0x7f48a7b22c6d - <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::InferCtxtPrivExt>::report_projection_error
  18:     0x7f48a7b16d94 - <rustc_infer[367541f3d053723]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[e87640d4ef5b721]::traits::error_reporting::TypeErrCtxtExt>::report_fulfillment_errors
  19:     0x7f48a5070188 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_argument_types
  20:     0x7f48a52483f9 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_call
  21:     0x7f48a51a23ce - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  22:     0x7f48a51f57a3 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_block_with_expected
  23:     0x7f48a51a27cf - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  24:     0x7f48a518deb3 - <rustc_hir_typeck[62f7ea54373b9441]::fn_ctxt::FnCtxt>::check_return_expr
  25:     0x7f48a518365f - rustc_hir_typeck[62f7ea54373b9441]::check::check_fn
  26:     0x7f48a516e47b - rustc_hir_typeck[62f7ea54373b9441]::typeck
  27:     0x7f48a554d4a2 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::VecCache<rustc_span[323ba4f8bef90746]::def_id::LocalDefId, &rustc_middle[962fafeac42f2169]::ty::context::TypeckResults>>
  28:     0x7f48a6636c05 - rustc_data_structures[13617afd267f24a6]::sync::par_for_each_in::<&[rustc_span[323ba4f8bef90746]::def_id::LocalDefId], <rustc_middle[962fafeac42f2169]::hir::map::Map>::par_body_owners<rustc_hir_typeck[62f7ea54373b9441]::typeck_item_bodies::{closure#0}>::{closure#0}>
  29:     0x7f48a6636943 - rustc_hir_typeck[62f7ea54373b9441]::typeck_item_bodies
  30:     0x7f48a65552d7 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::DefaultCache<(), ()>>
  31:     0x7f48a6554fe0 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::get_query::<rustc_query_impl[3365ff5720cc62]::queries::typeck_item_bodies, rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt>
  32:     0x7f48a52e2c02 - <rustc_session[f586d2b4d440cc8a]::session::Session>::time::<(), rustc_hir_analysis[4b631a2ab94469f]::check_crate::{closure#7}>
  33:     0x7f48a52e19dc - rustc_hir_analysis[4b631a2ab94469f]::check_crate
  34:     0x7f48a52e15fb - rustc_interface[e2cb08752aecdf39]::passes::analysis
  35:     0x7f48a669ec0a - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::try_execute_query::<rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt, rustc_query_system[a940e1e33b00a8d3]::query::caches::DefaultCache<(), core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>>
  36:     0x7f48a669e900 - rustc_query_system[a940e1e33b00a8d3]::query::plumbing::get_query::<rustc_query_impl[3365ff5720cc62]::queries::analysis, rustc_query_impl[3365ff5720cc62]::plumbing::QueryCtxt>
  37:     0x7f48a603a2ab - <rustc_interface[e2cb08752aecdf39]::passes::QueryContext>::enter::<rustc_driver[b218747198152f07]::run_compiler::{closure#1}::{closure#2}::{closure#2}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  38:     0x7f48a6031c5e - <rustc_interface[e2cb08752aecdf39]::interface::Compiler>::enter::<rustc_driver[b218747198152f07]::run_compiler::{closure#1}::{closure#2}, core[e1ff35b71d1701d9]::result::Result<core[e1ff35b71d1701d9]::option::Option<rustc_interface[e2cb08752aecdf39]::queries::Linker>, rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  39:     0x7f48a602cc78 - rustc_span[323ba4f8bef90746]::with_source_map::<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  40:     0x7f48a602c765 - <scoped_tls[b7ce8b7e4286e02d]::ScopedKey<rustc_span[323ba4f8bef90746]::SessionGlobals>>::set::<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  41:     0x7f48a602bd52 - std[2f49e04a49785d41]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e2cb08752aecdf39]::util::run_in_thread_pool_with_globals<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>
  42:     0x7f48a67a5458 - <<std[2f49e04a49785d41]::thread::Builder>::spawn_unchecked_<rustc_interface[e2cb08752aecdf39]::util::run_in_thread_pool_with_globals<rustc_interface[e2cb08752aecdf39]::interface::run_compiler<core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>, rustc_driver[b218747198152f07]::run_compiler::{closure#1}>::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[e1ff35b71d1701d9]::result::Result<(), rustc_errors[3f04fd9a36ec0931]::ErrorGuaranteed>>::{closure#1} as core[e1ff35b71d1701d9]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  43:     0x7f48a7cb5df3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5d679a3e0eef548a
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/alloc/src/boxed.rs:2000:9
  44:     0x7f48a7cb5df3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5129455b92134839
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/alloc/src/boxed.rs:2000:9
  45:     0x7f48a7cb5df3 - std::sys::unix::thread::Thread::new::thread_start::haa6aecc225cb6cab
                               at /rustc/53e4b9dd74c29cc9308b8d0f10facac70bb101a7/library/std/src/sys/unix/thread.rs:108:17
  46:     0x7f48a39068fd - <unknown>
  47:     0x7f48a3988a60 - <unknown>
  48:                0x0 - <unknown>
thread panicked while panicking. aborting.
```

</p>
</details>

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -1607,7 +1607,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         | ObligationCauseCode::ObjectCastObligation(..)
                         | ObligationCauseCode::OpaqueType
                 );
-                let expected_ty = data.term.ty().unwrap();
+                let expected_ty = data.term.ty().unwrap_or_else(|| self.tcx.ty_error());
 
                 // constrain inference variables a bit more to nested obligations from normalize so
                 // we can have more helpful errors.

--- a/src/test/ui/issues/issue-105330.rs
+++ b/src/test/ui/issues/issue-105330.rs
@@ -1,0 +1,21 @@
+pub trait TraitWAssocConst {
+    const A:   usize;
+}
+pub struct Demo {}
+
+impl TraitWAssocConst for impl Demo { //~ ERROR E0404
+    //~^ ERROR E0562
+    pubconst A: str = 32; //~ ERROR expected one of
+}
+
+fn foo<A: TraitWAssocConst<A=32>>() { //~ ERROR E0658
+    foo::<Demo>()(); //~ ERROR E0271
+    //~^ ERROR E0618
+    //~| ERROR E0277
+}
+
+fn main<A: TraitWAssocConst<A=32>>() { //~ ERROR E0131
+    //~^ ERROR E0658
+    foo::<Demo>(); //~ ERROR E0277
+    //~^ ERROR E0271
+}

--- a/src/test/ui/issues/issue-105330.stderr
+++ b/src/test/ui/issues/issue-105330.stderr
@@ -1,0 +1,109 @@
+error: expected one of `!` or `::`, found `A`
+  --> $DIR/issue-105330.rs:8:14
+   |
+LL | impl TraitWAssocConst for impl Demo {
+   |                                     - while parsing this item list starting here
+LL |
+LL |     pubconst A: str = 32;
+   |              ^ expected one of `!` or `::`
+LL | }
+   | - the item list ends here
+
+error[E0404]: expected trait, found struct `Demo`
+  --> $DIR/issue-105330.rs:6:32
+   |
+LL | impl TraitWAssocConst for impl Demo {
+   |                                ^^^^ not a trait
+
+error[E0658]: associated const equality is incomplete
+  --> $DIR/issue-105330.rs:11:28
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   |                            ^^^^
+   |
+   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
+   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable
+
+error[E0658]: associated const equality is incomplete
+  --> $DIR/issue-105330.rs:17:29
+   |
+LL | fn main<A: TraitWAssocConst<A=32>>() {
+   |                             ^^^^
+   |
+   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
+   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable
+
+error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in type
+  --> $DIR/issue-105330.rs:6:27
+   |
+LL | impl TraitWAssocConst for impl Demo {
+   |                           ^^^^^^^^^
+
+error[E0277]: the trait bound `Demo: TraitWAssocConst` is not satisfied
+  --> $DIR/issue-105330.rs:12:11
+   |
+LL |     foo::<Demo>()();
+   |           ^^^^ the trait `TraitWAssocConst` is not implemented for `Demo`
+   |
+note: required by a bound in `foo`
+  --> $DIR/issue-105330.rs:11:11
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   |           ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `foo`
+
+error[E0271]: type mismatch resolving `<Demo as TraitWAssocConst>::A == 32`
+  --> $DIR/issue-105330.rs:12:11
+   |
+LL |     foo::<Demo>()();
+   |           ^^^^ types differ
+   |
+note: required by a bound in `foo`
+  --> $DIR/issue-105330.rs:11:28
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   |                            ^^^^ required by this bound in `foo`
+
+error[E0618]: expected function, found `()`
+  --> $DIR/issue-105330.rs:12:5
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   | ----------------------------------- `foo::<Demo>` defined here returns `()`
+LL |     foo::<Demo>()();
+   |     ^^^^^^^^^^^^^--
+   |     |
+   |     call expression requires function
+
+error[E0277]: the trait bound `Demo: TraitWAssocConst` is not satisfied
+  --> $DIR/issue-105330.rs:19:11
+   |
+LL |     foo::<Demo>();
+   |           ^^^^ the trait `TraitWAssocConst` is not implemented for `Demo`
+   |
+note: required by a bound in `foo`
+  --> $DIR/issue-105330.rs:11:11
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   |           ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `foo`
+
+error[E0271]: type mismatch resolving `<Demo as TraitWAssocConst>::A == 32`
+  --> $DIR/issue-105330.rs:19:11
+   |
+LL |     foo::<Demo>();
+   |           ^^^^ types differ
+   |
+note: required by a bound in `foo`
+  --> $DIR/issue-105330.rs:11:28
+   |
+LL | fn foo<A: TraitWAssocConst<A=32>>() {
+   |                            ^^^^ required by this bound in `foo`
+
+error[E0131]: `main` function is not allowed to have generic parameters
+  --> $DIR/issue-105330.rs:17:8
+   |
+LL | fn main<A: TraitWAssocConst<A=32>>() {
+   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `main` cannot have generic parameters
+
+error: aborting due to 11 previous errors
+
+Some errors have detailed explanations: E0131, E0271, E0277, E0404, E0562, E0618, E0658.
+For more information about an error, try `rustc --explain E0131`.

